WIP: Reports module

tsalo opened the pull request:
This PR adds the reports module for generating the data acquisition portion of a methods section from BIDS datasets. There's a long way to go, but it's an improvement on the example code I linked to in #99.

tyarkoni commented:
I'll review this more carefully in the next few days. One general thought: rather than just having the `task_converter` passed in to `.generate()`, maybe allow the user to pass it to `BIDSReport.__init__`. It could also be a hierarchical dict, so that the first-level keys are `'task'`, `'dir'`, `'seqvar'`, etc. Basically, the user should have the option to override any of the built-in dicts you've already defined.

Then, for methods like `.generate()`, you can still have an optional dict passed in as well--this one just for task name substitutions. Then, internally, the logic is that when a `BIDSReport` is initialized, you update the default package-level mapping dictionary with anything passed to `init` (optionally). And then for `generate()` or other similar calls, you update again with any optional arguments. So you allow users to override your default mappings both at report initialization, and at generation-time.

Another benefit of structuring it this way is you can follow the grabbids approach of having a JSON file containing all the default configuration information as a separate file in the module. I.e., move all the `DIR_CONVERTER`, `SEQ_CONVERTER`, etc. dicts out of the code and into a JSON config file.

tsalo commented:
That makes sense to me. I just made a commit (047136d3e33854b24f82efbfcefd2428986fcbd4) that should do that. Now, there's a json for the hierarchical dict in the config subfolder. `BIDSReport.__init__` has an optional `config` argument (if `None` it grabs the json), and the `task_converter` argument from `BIDSReport.generate` will just overwrite `BIDSReport.config['task']`.

tsalo commented:
I'm planning to merge some changes from @emdupre soon. She's working on making the reports compatible with multi-echo datasets. I've added the README and have resolved the remaining conflicts, so once the multi-echo stuff is done, I think it should be ready to merge.

tyarkoni commented:
Cool, no rush at all. Just let me know when you're ready, and I'll do a full review.

tsalo commented:
Sorry for the delay. I didn't get around to merging @emdupre's PR until now, but everything looks good and it's ready for review.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+            with open(config) as fobj:
+                config = json.load(fobj)
+
+        if not isinstance(config, dict):
```
I think you're missing the case where `config` is not `None`, but a filename is passed... this will fail on the current check.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+
+        Parameters
+        ----------
+        task_converter : :obj:`dict`
```
I know it's a bit obnoxious to ask for better docs when I'm not good at doing that myself, but I'd suggest either adding an `Example` section, or giving a bit more information about what the keys and values in the `task_converter` dict are.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+            A dictionary with information for converting task names from BIDS
+            filename format to human readable strings.
+        """
+        if task_converter is None:
```
Should the default conversion dictionary also move to `converters.json`?

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+
+        self.config = config
+
+    def generate(self, task_converter=None):
```
Is there a reason to pass the `task_converter` here, but all the other converters in the initializer? I get that tasks might vary much more on a project-by-project basis than most of the other fields, but since the user has to pass a single `BIDSLayout` to the init, the expectation is presumably that any `generate` calls are always going to apply to the same project. So it might make sense to move `task_converter` to the initializer as well--unless you think it will be a common use case for someone to only want to do one task out of many, and specify the task dict on the fly instead of creating one for the project as a whole.

tsalo replied:
I was thinking it might be good just to use the `TaskName` field in the metadata, when available. Presumably, that would be the same full name as the one users would pass in as `task_converter`. Otherwise, I can move it back into `__init__` and just merge `task_converter` into `config` at the beginning.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+        if task_converter is None:
+            task_converter = {'rest': 'resting state'}
+
+        self.config['task'] = task_converter
```
See above comment about `generate`: if this is something that should be stored globally in the `config`, I would probably move the argument to `init`. If it isn't, then I don't think we need to update `self.config` here.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+
+        descriptions = []
+
+        subjs = self.layout.get_subjects()
```
You might want to allow `**kwargs` to be passed through to `get_subjects`, as the user might want to apply constraints (e.g., `task` might be a common one).

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+        print(utils.warnings())
+        return counter
+
+    def _report(self, subj):
```
Maybe call this `_report_subject` or `_subject_report`, as I can imagine there may be other private report methods later.

tsalo replied:
Good point. Will do.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+            descriptions.append(description)
+        counter = Counter(descriptions)
+        print('Number of patterns detected: {0}'.format(len(counter.keys())))
+        print(utils.warnings())
```
This is a bit pedantic, but maybe call this `reminder` or something. If it's at the level of a legit warning, it should probably be issued via `warnings.warn` rather than printed.

tsalo replied:
Yeah.. that wasn't the best naming on my part. Reminder makes more sense.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+            information. Each scan type is given its own paragraph.
+        """
+        description_list = []
+        sessions = self.layout.get_sessions()
```
This should be constrained by passing `subject=subj`, as the available sessions might differ by subject

tsalo replied:
Good point. Will do.

tyarkoni commented on bids/reports/utils.py:
```diff
@@ -0,0 +1,573 @@
+        List of nifti files in layout corresponding to subject/session combo.
+    subj : :obj:`str`
+        Subject ID.
+    ses : :obj:`str` or :obj:`None`
```
If this can be `None`, set that as the default value in the argument list. This should also make your life easier when handling sessions as I suggested above.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+        """
+        description_list = []
+        sessions = self.layout.get_sessions()
+        if sessions:
```
I think you can simplify the next ~10 lines by requesting the niftis without the `session` constraint and asking for `return_type='obj'` or `return_type='tuple'`. Then you can inspect the returned object to see if it has a `.session` set (or, for `'obj'`, it will have a key in `.entities`), and add anything you need (and from the looks of it, you could potentially just pass it through to `utils.parse_niftis` even if it's `None`).

tsalo replied:
I'll try doing this, but would it be possible for `BIDSLayout.get` to treat keyword arguments with `None` as if they weren't included? That would make all of this a lot easier. I'm not familiar with `grabbit`, though, so I have no clue if it's feasible.

tyarkoni replied:
No, the `get` call can't ignore args with value `None`, but you can probably consolidate all the arguments in kwargs before passing, and then do something like `kwargs = {k: v for k, v in kwargs.items() if v is not None}

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+            niftis = self.layout.get(subject=subj, extensions='nii.gz')
+
+            if niftis:
+                description_list += utils.parse_niftis(self.layout, niftis, subj, None, self.config)
```
See other comments, but it's probably better to extract the value of `session` (including possibly `None` and then pass it like the other args.

tyarkoni commented on bids/reports/report.py:
```diff
@@ -0,0 +1,115 @@
+
+        # Assume all data were converted the same way and use the first nifti file's
+        # json for conversion information.
+        if 'metadata' not in vars():
```
If this is just the final paragraph? Is it essential that we fail? I think if the user can get back *some* usable information, it might be worth issuing a `warning` instead, and still returning `description`. But I haven't thought about it, so maybe this doesn't make sense.

tsalo replied:
This should raise an exception if no metadata could be found for any of the scans in the dataset. At least for now, if that exception is raised then description is empty anyway.

When there are only niftis, the only information that could be extracted are the number of slices, matrix size, field of view, and voxel size. At least at the moment, the parsing functions won't return that information if metadata couldn't be found, so nothing would be returned. If we end up making those functions more flexible, then I guess very short descriptions could be created and returned for the scans, but I don't know if people will use the module in cases where so little information can be extracted.

tyarkoni replied:
Ah, that makes sense.

tyarkoni commented on bids/reports/utils.py:
```diff
@@ -0,0 +1,573 @@
+    return str_
+
+
+def get_slice_info(slice_times):
```
I think the `utils` module is a bit big and contains a bunch of functions that feel to me like they're doing core parsing/extraction stuff. It might make sense to keep `utils` for things like `list_to_str`, and put heavier functions like `get_slice_info` and `get_seqstr` into their own module called something more descriptive like `format` or `parsing` or something. But it's your call.

tsalo replied:
Good idea. I'll move all of the string formatters into `parsing` and keep the actual utilities in `utils`. For now, I'm thinking of keeping `get_slice_info`, `get_seqstr`, and `get_sizestr` in `utils`, mainly because `parsing` still ends up being pretty big even when it only contains the `*_info` functions.

tyarkoni commented on bids/reports/utils.py:
```diff
@@ -0,0 +1,573 @@
+        Output string with scanner information.
+    """
+    if 'ConversionSoftware' in metadata.keys():
+        software_str = ' using {soft} ({conv_vers})'.format(soft=metadata['ConversionSoftware'],
```
Indentation looks off (though it could just be a GitHub issue); maybe run this all through a linter if you haven't already, and fix any glaring PEP8 issues.

tsalo commented:
Thanks for the feedback. I'm working on addressing your comments, but am having some trouble with writings tests. The only dataset in `tests/data` with actual data in the nifti files is `synthetic`, but it lacks some metadata that the module currently requires (e.g., `MagneticFieldStrength`). I can either make the code more flexible or we can update the test data, depending on what information is really necessary for publication.

I came up with a list of all necessary and optional information used to generate the reports, split by scan type. Maybe you could help figure out which ones should actually be required (and potentially be accounted for in the `bids-validator`) and which ones the module should work around?

Here they are:
- Used for all scans:
  - MagneticFieldStrength
  - Manufacturer
  - ManufacturersModelName
  - RepetitionTime
  - EchoTime
  - FlipAngle
  - ScanningSequence (optional)
  - SequenceVariant (optional)
  - Nifti data (arg)
    - Needed to determine matrix size, field of view, voxel size, and number of TRs/directions.
- Functional scans:
  - SliceTiming
  - Number of runs (arg)
  - Task name (arg)
- Structural scans:
  - None
- DWI scans:
  - Bval file (arg)
- Field maps:
  - PhaseEncodingDirection
  - IntendedFor (optional)

tyarkoni commented:
Re: tests, I think it's probably most important to test the individual formatting methods, and those shouldn't need real data. I.e., you can just create fake data (feel free to put it in `reports/tests/data` if they're too big to embed in the test file) that you just feed directly to the helpers for parsing. It's not essential that the input strings come from a real dataset, provided they're realistic.

tsalo commented:
I added a test for each function in `reports.parsing`, although, since they return long strings, the tests just make sure that the functions pass and return the right data types.

tyarkoni commented:
@tsalo, I restarted the tests following some merged PRs; assuming they pass, is this good to merge?

tyarkoni commented:
@tsalo when you get a chance, can you merge master into this branch? The tests are failing and I suspect updating your branch will fix them. Thanks!

tsalo commented:
I updated from master and I think it's good to merge if the tests pass. Fingers crossed!

tsalo commented:
The appveyor failures seem to be caused by tests in `analysis` and `variables`, but not `reports`, although I'm not sure why they're breaking.

tyarkoni commented:
Yeah, I'm aware of those (see #139); we'll ignore them for now. Merging this--thanks!
